Overhaul multicaret editing and selection in TextEdit

kitbdev opened the pull request:
- related #73471
- fixes #86863 (all issues except some of [reload script](https://github.com/godotengine/godot/issues/86863#reload-scripts) and [goto line](https://github.com/godotengine/godot/issues/86863#goto-line))
- fixes #83410
- fixes #72797
- probably fixes #74400 (can't reproduce, related to other issues fixed)
- fixes #73202
- fixes #81535
- fixes #83826
- fixes https://github.com/godotengine/godot/issues/80378
- fixes #83589
- fixes https://github.com/godotengine/godot/issues/90462
- supersedes (can still cherry-pick) #84901 
- supersedes (can still cherry-pick) #81331
- supersedes #83451

Selection now goes from selection origin to the caret, so that the caret is always attached to the selection.

All TextEdit APIs now handle multiple carets. This should make it easier to use and use correctly.
If a method is called alone, it will merge overlapping carets. If it is called in a loop for multiple carets as part of a multicaret edit, the merge will happen afterwards to not affect caret indexes.

All text changes and caret changes now `queue_redraw()`.

<details>
<summary>Implementation details</summary>

General
- Added `remove_line_at()` and `insert_text()` in TextEdit to flesh out the API because methods now affects carets in different ways, so the way it is used and the context matters. 
	- For example: `set_line()` was often used to insert text, but this doesn't allow for much context of how the carets on that line should be affected.
- Moved `move_lines_up/down()`, `delete_lines()`, `duplicate_selection()`, and `delete_lines()` from `CodeEditor` to `CodeEdit` since the implementations changed so they can be tested, and based on this https://github.com/godotengine/godot/pull/66553#pullrequestreview-1598392668. 
- Made drag and drop text use a new caret instead of the main one so it won't interfere with any carets.
	- `get_caret_count()` and many other methods now ignores the drag caret if it exists, as it shouldn't be used as a regular caret in most cases.
	- I wanted it to not be part of the `carets` vector at all, but I also didn't want to copy lots of drawing and scrolling code.
- Added `_text_changed()`, `_caret_changed()`, and `_selection_changed()` to have one clear place after a change that will be called immediately. These are used to redraw, defer emit signals, and cancel drag and drop if needed.
	- A `selection_changed` signal should be easy to add now, if it is needed in the future.
- Added `_cancel_drag_and_drop_text()` to cancel the Viewport's drag and drop in certain circumstances.
	- Added `Viewport::gui_cancel_drag()` to support this, since I couldn't find an existing way of cancelling a drag and drop operation.
- Added `_unhide_carets()` because sometimes the TextEdit needs to unfold a line, and it can't just use unhide the line.
- Changed `_cut_internal()` to use `_copy_internal()` then just remove the selection/lines as it is much simpler.
- In `set_caret_line()` the `p_wrap_index` param now accepts `-1` to only clamp the column and not try to adjust it. In many cases `set_caret_column()` is called immediately afterward, so calculating the column so it lines up visually was unnecessary, and sometimes needed a workaround to avoid.
- Changed some MessageQueue calls to use callables, since I moved them around. (related #86301)
- Added `cancel_ime()` and `apply_ime()` for more control over the IME.
	- Added internally `_close_ime_window()`, `_update_ime_window_position()`, and `_update_ime_text()` to reduce code duplication and increase clarity.
	- We may want to make similar changes in LineEdit?
- Added `get/set_carets_state()` for the various `reload_text()` methods and for getting/setting the entire state.
- Updated some comments style. I didn't change the block comments that are used as headers, since they should probably be distinct from regular comments and I didn't know what style to change them to.
- Changed `GotoLineDialog::popup_find_line()` to use a `CodeTextEditor` instead of a `CodeEdit` so it can use the existing `goto_line()` functionality and not need to recreate it. 
	- GotoLineDialog should use a SpinBox or something, but that can be a separate PR.

TextEdit Multicaret
- Added `begin/end_multicaret_edit()` to be able to allow merging the carets when the edit is complete.
	- These should always come in pairs.
	- These can be nested, and is what allows for more complex editing to use regular APIs, and have both merge only once at the end of the edit.
- Added `queue_merge_carets()` to be used when merging should happen at the end of the multicaret edit.
- Added `multicaret_edit_ignore_caret()` that should be used when iterating over carets when editing them and text is removed. This will return true when the given caret would have been removed due to it merging with another caret, or it was just added. 
	- This allows multiple carets to be collapsed into one and delay the merging without affecting the edit.
- Added `collapse_carets()` to move all carets in a region of deleted text to one overlapping location and add to the ignore list. These will be merged at the end of the edit unless they are separated.
- Removed `get_caret_index_edit_order()`, because it is no longer needed since carets can be edited in any order.
- Added `get_sorted_carets()` to replace some functionality from `get_caret_index_edit_order`, but in the opposite order. It sorts by selection_from_line instead of selection_to_line, in ascending order (top of page to bottom).
- `get_sorted_carets()` is const and doesn't cache because it is usually only needed after a caret is moved which would invalidate the cache. I also use a Comparator so it should be faster, O(log(n)) instead of O(n^2). If there are still performance concerns, we can change it to use a cache.
- Added `get_line_ranges_from_carets()` to get a vector of all the lines that are part of a selection or have a caret on them. This is useful for many common editing tasks.
	- This was inspired by the removed methods in CodeTextEditor `_get_affected_lines_from/to()`, but more comprehensive.
- Removed `adjust_carets_after_edit()` because it is no longer needed now that carets are automatically handled. 
	- `_offset_carets_after()` is similar (though not the same) and is used internally for a similar purpose, to move all carets after some text was added/removed.
- Updated all code that operates on multiple carets.
	- `TextEdit::merge_overlapping_carets()`, `TextEdit::_new_line()`, `TextEdit::_do_backspace()`, `TextEdit::_delete()`, `TextEdit::insert_text_at_caret()`, `TextEdit::add_caret_at_carets()`, `TextEdit::adjust_carets_after_edit()`, `TextEdit::get_selected_text()`, `TextEdit::delete_selection()`, `TextEdit::_handle_unicode_input_internal()`, `TextEdit::_backspace_internal()`, `TextEdit::_cut_internal()`, `TextEdit::_copy_internal()`, `TextEdit::_paste_internal()`, `CodeEdit::_handle_unicode_input_internal()`, `CodeEdit::_backspace_internal()`, `CodeEdit::do_indent()`, `CodeEdit::indent_lines()`, `CodeEdit::unindent_lines()`, `CodeEdit::_new_line()`, `CodeEdit::create_code_region()`, `CodeEdit::confirm_code_completion()`, `CodeEdit::duplicate_lines()`, `TextEditor::_edit_option(EDIT_TOGGLE_FOLD_LINE)`, `ScriptTextEditor::_edit_option(EDIT_TOGGLE_FOLD_LINE)`, `ScriptTextEditor::_edit_option(DEBUG_TOGGLE_BREAKPOINT)`, `CodeTextEditor::convert_case()`, `CodeTextEditor::move_lines_up()`, `CodeTextEditor::move_lines_down()`, `CodeTextEditor::delete_lines()`, `CodeTextEditor::duplicate_selection()`, `CodeTextEditor::toggle_inline_comment()`, `CodeTextEditor::toggle_bookmark()`

TextEdit Selection
- Changed `select()` to now select from a 'selection origin' to the caret. Calling it will move the caret.
- `get_selection_from/to_line/column` still gives the start/end of the selection and can be used like before. Since these values are now calculated and its a common use case, they return the caret line/column if there is no active selection.
- Added `set_selection_origin_line/column()` to be able to set the origin of the selection directly, similar to `set_caret_line/column()`.
	- Added a `get_selection_origin_line/column()`. Deprecated `get_selection_line/column()` because the functionality is a bit different, and I wanted a more clear name. It didn't always return the origin of the selection, it was just used to setup the different selection modes.
	- These functions work even if there is no selection because it is sometimes needed to use or set it before starting a selection, like in `_pre_shift_selection()` or `add_caret_at_carets()`.
- Selections can now touch each other since when I was reworking on `merge_overlapping_carets()`, I used VSCode for reference and that is how it is done there. It is also useful in some scenarios.
- Changed `has_selection()`, `get_selected_text()`, and `deselect()` to be faster O(1) if the caret index is passed in instead of O(n). I didn't do this for all of them since there would be lots of copy pasted code.
- Added `is_selection_direction_right()` to be able to tell if the selection is left to right (top to bottom) or the reverse, since it now always has a direction.
- Removed `_post_shift_selection()` as it wasn't needed anymore.
- This comment in `_click_selection_held()` is no longer true: 
	- `Warning: is_mouse_button_pressed(MouseButton::LEFT) returns false for double+ clicks...`
	- I'm not sure when this changed but it does work for word and line mode.
- Added `selection_contains()`, `get_selection_at()`, and `is_line_col_in_range()` as helper functions to reduce duplicate code.

Tests
- Added clipboard and primary clipboard support to DisplayServerMock to be able to test cut, copy, paste, and paste primary
	- Removed those todos
- Added `build_array()` to TestTextEdit and TestCodeEdit to be able to easily create arrays, copied from `variant/test_array.h`.
	- These can be replaced with initializer lists after #86015.
- Updated comment style for CodeEditTests.
- Added `SEND_GUI_KEY_UP_EVENT` test macro, for some text drag tests where the key up event was needed.
- Added test cases for `insert text`, `remove text`, `remove line at`
- Improved/added tests in areas that had bugs or where I changed lots of code.
- Removed todo `// Add undo / redo tests?` line 1368 by adding undo redo tests for each input subcase. 
- Lots of `lines_edited_args` had to be changed since the order edits are done in are different, but this shouldn't be an issue.
- Added some comments to clarify some tests.
- Added `// FIXME: Remove after GH-77101 is fixed.` to some input tests, since a new action should not need to be started in between edits with different carets. related #77101.
- Removed `caret index edit order` tests, Added `sort carets` and `merge carets` tests.
- Added `text manipulation` test case with subcases `backspace`, `new line`,`move lines up`,`move lines down`,`delete lines`,`duplicate selection`,`duplicate lines`.

</details>

AThousandShips commented:
You're breaking compatibility and need to implement compatibility methods, if you need help I can instruct when I have the time 🙂

AThousandShips commented on scene/gui/text_edit.h:
```diff
@@ -838,23 +843,33 @@ class TextEdit : public Control {
 
-	int add_caret(int p_line, int p_col);
+	Dictionary get_carets_state() const;
+	void set_carets_state(Dictionary p_caret_state);
```
```suggestion
	void set_carets_state(const Dictionary &p_caret_state);
```
Pass by reference

AThousandShips commented on doc/classes/TextEdit.xml:
```diff
@@ -83,15 +83,15 @@
 			<param index="4" name="to_col" type="int" />
 			<description>
-				Reposition the carets affected by the edit. This assumes edits are applied in edit order, see [method get_caret_index_edit_order].
+				[i]Obsoleted.[/i] No longer necessary since methods now adjust carets themselves.
```
```suggestion
				[i]Deprecated.[/i] No longer necessary since methods now adjust carets themselves.
```
Agreed upon syntax in new content, and we use the same wording regardless if it's now non-functional or just slated for future removal

AThousandShips commented on doc/classes/TextEdit.xml:
```diff
@@ -502,32 +514,32 @@
 			<param index="0" name="caret_index" type="int" default="0" />
 			<description>
-				Returns the original start column of the selection.
+				[i]Obsoleted.[/i] Use [method get_selection_origin_column] instead.
```
```suggestion
				[i]Deprecated.[/i] Use [method get_selection_origin_column] instead.
```

AThousandShips commented on doc/classes/TextEdit.xml:
```diff
@@ -502,32 +514,32 @@
 			<param index="0" name="caret_index" type="int" default="0" />
 			<description>
-				Returns the original start line of the selection.
+				[i]Obsoleted.[/i] Use [method get_selection_origin_line] instead.
```
```suggestion
				[i]Deprecated.[/i] Use [method get_selection_origin_line] instead.
```

AThousandShips commented:
Many of the comment changes here are in unrelated areas of code and while appreciated they do complicate the git history, I think it should be limited to areas where you're already changing things to avoid chaos

kitbdev commented:
Added compat file and removed comment changes.

MewPurPur commented on doc/classes/CodeEdit.xml:
```diff
@@ -146,6 +146,12 @@
+		<method name="delete_lines">
+			<return type="void" />
+			<description>
+				Deletes all lines that are selected or have a caret on it.
```
it -> them

The mistake is made multiple times; I won't spam review comments about it.

MewPurPur commented on doc/classes/TextEdit.xml:
```diff
@@ -114,6 +114,12 @@
+		<method name="begin_multicaret_edit">
+			<return type="void" />
+			<description>
+				Starts an edit for multiple carets. The edit must be ended with [method end_multicaret_edit]. Multicaret edits can be used to edit text at multiple carets and delay merging the carets until the end, so the caret indexes aren't affected immediately. [method begin_multicaret_edit] and [method end_multicaret_edit] can be nested, and the merge will happen at the last [method end_multicaret_edit].
```
Is there any point in nesting? This seems like just a toggle between "no caret merging mode" and "caret merging mode" to me.

kitbdev replied:
Yes, since some methods like `insert_text_at_caret()` and `delete_selection()` need to be able to operate on all carets and on just one as part of another edit, so it's useful to be able to nest them.
While I don't expect users to need to call functions that operate on multiple carets within a loop, they still might. Also there's no reason to not have it, since we still need to know when a multicaret edit starts and ends.

I could remove the nesting part of the docs, to not encourage it?

Paulb23 reviewed (commented):
It's been a while, always enjoy your PRs :)

Haven't had a chance to get fully into the details here, but I like the direction.
Given the range / scope and size of changes here, probably best not to try and do this in one go. Is there any chance you could spilt some of these change up into multiple PR / commits?

Otherwise a few notes on an initial look: 
> In `set_caret_line()` the p_wrap_index param now accepts -1 to only clamp the column

Have thought about adding a `caret_set_pos(line, col)` to get around some of this API awkwardness.

> Added `queue_merge_carets()` to be used when merging should happen at the end of the multicaret edit.

If we're trying to simply the interface could we always call `merge_overlapping_carets`? We kind of do this for  `*_move_caret_*` methods, then the API users don't have to care about detecting overlaps? Given the caret ignoring / merging should now be taken care of for them via `collapse_carets` (should this be exposed?).

> O(1) if the caret index is passed in instead of O(n). I didn't do this for all of them since there would be lots of copy pasted code.

Yeah the only way I was thinking is to create a bunch of internal methods. But I can't see users using that many carets that it becomes an immediate issue.

> Removed get_caret_index_edit_order(), because it is no longer needed since carets can be edited in any order.

The intention behind this was to make editing consistent regardless of order the carets where placed down. Plus processing carets twice. I guess due to delaying the merge / ignore carets this becomes a non-issue?

> begin/end_multicaret_edit() vs begin/end_complex_operation()

Would we want to force all multicaret edits to be complex operations? Otherwise the undo / redo will not be great. Though you might want a complex op to consist of multiple multicaret edits, so don't think we can combine them entirely.

> `is_selection_direction_right`

I do wonder if there's another way to name this because with `right-to-left`, right is not always right.

kitbdev commented:
> Is there any chance you could spilt some of these change up into multiple PR / commits?

This is tricky since the multicaret edit change and the selection change require lots of functions to be rewritten, and some of them depend on each other.
I think I can split out some smaller parts into separate PRs, like the queue_redraw change and the IME changes. This PR would probably depend on them.

>If we're trying to simply the interface could we always call merge_overlapping_carets? 

I was thinking that users may still want to force a merge during a multicaret edit, but I'm not sure if there is a use case for that so I think we can remove `queue_merge_carets()` and put the functionality in `merge_overlapping_carets()`.

>via collapse_carets (should this be exposed?)

It doesn't need to be exposed since all methods that remove text call it internally. It's only public because it's useful for `CodeEdit::fold_line()`. If there are more uses for it like that, then we can expose it.

> The intention behind this was to make editing consistent regardless of order the carets where placed down. Plus processing carets twice. I guess due to delaying the merge / ignore carets this becomes a non-issue?

Yes, with the delayed merge, ignored carets, and removing text updating carets after it, the order no longer has to be from bottom to top and can be in any order.

> Would we want to force all multicaret edits to be complex operations? Otherwise the undo / redo will not be great. Though you might want a complex op to consist of multiple multicaret edits, so don't think we can combine them entirely.

There are also some places where we want a multicaret edit to not be part of undo / redo at all, like `add_caret_at_carets`, so we can't combine them.

> is_selection_direction_right -> is_caret_after_selection_origin ?
`is_caret_after_selection_origin` is more

kitbdev commented:
Fixed merge conflict, combined `queue_merge_carets()` and `merge_overlapping_carets()`, and renamed `is_selection_direction_right` to `is_caret_after_selection_origin`.

I'm going to try and extract some smaller PRs that can be merged more easily.

MewPurPur commented:
I'm testing this PR right now to help get it through. Should operations always have the same effect, regardless of the order in which you created carets?

MewPurPur commented:
I'll assume yes to my above post and report my findings, I don't wanna delay them. In general, I've found very few new issues.

```
lmofa
|lmo|fa
```

The result after pressing ui_text_backspace_word (Ctrl+Backspace) or ui_text_backspace_all_to_left (no default shortcut) is different depending on the order of placing down the carets.

fun|c lmofa():|
  pa|ss


The result after pressing Fold/Unfold Line (Alt+F) is an error in output.

-----

Ctrl+Shift+U for inputting unicode has several new bugs. On success, Ctrl+Shift+U no longer deletes the unicode stuff after you're done typing. With multicaret, carets easily get misplaced. **If you do it with a selection, Godot pretty much breaks.**

kitbdev commented:
The backspace issue also happened with ui_text_delete_word when one of the carets was at the end of the line and the other on the last word. It doesn't apply to ui_text_delete_all_to_right though since it doesn't delete newlines.
I fixed the issue by making sure newlines are handled after removing text for the line, and added tests for it.
Fixed the fold line issue and added a test for it.

I couldn't find what you meant for the unicode issue. For me `Ctrl+Shift+U` isn't bound to anything and I didn't see any actions or functions about unicode that would function like that.
The only unicode input I know is holding ALT and typing `+` and the code, like `+1F600`. But it seems to be working fine.

MewPurPur commented:
Oh, so maybe Ctrl+Shift+U is linux-specific? I was discussing it with a friend who's on Windows once, I thought it's universal.

Anyway, when you do this shortcut, an underlined "u" appears, and then you write a 4-letter code and get a unicode character when you press Enter.

This worked before somewhat ok, now it doesn't work.

Paulb23 commented:
> I think I can split out some smaller parts into separate PRs, like the queue_redraw change and the IME changes. This PR would probably depend on them.

As many as you feel happy to would be appreciated, dependent PRs are fine as they'll hopefully be smaller so can be merged quicker (with a little reabasing), or even creating separate commits would be okay! 

> If there are more uses for it like that, then we can expose it.

I'm leaning towards adding any public method to the API by default so we don't have too much "editor" only code if we can help it. As if we're making it public then there's a need for it somewhere. If there's really no use for it being called outside `TextEdit / CodeEdit` though, protected is fine (particularly around folding as that is awkwardly split at the moment).

> There are also some places where we want a multicaret edit to not be part of undo / redo at all, like `add_caret_at_carets`, so we can't combine them.

This seems a little odd to me, as adding / removing carets shouldn't be changing any text? Looking at  `add_caret_at_carets` in particular, it calls `merge_overlapping_carets` at the end which should achieve the same purpose?

kitbdev committed:
Overhaul TextEdit selection.
The caret is now a part of the selection.

kitbdev committed:
Overhaul multiple caret editing in TextEdit.
Use a multicaret edit to delay merging overlapping carets until the end.

kitbdev commented:
I split the commit into two parts:
- Selection overhaul - includes changes to selection, mouse handling, and goto line changes.
- Multicaret edit overhaul - includes multicaret edits, everything else.

It's not perfect since I don't think the first commit could be built by itself and most things are just in the last commit, but hopefully it helps with reviewing the code.
I can remerge them later, if we want.

> I'm leaning towards adding any public method to the API by default 

In this case, I will make `_is_line_col_in_range` and `_selection_contains` private, since they aren't needed elsewhere, and I'll expose `get_selection_at` and `collapse_carets`.
I wasn't able to expose `get_line_ranges_from_carets`, since I got compiler errors, I guess `Vector<Point2i>` can't be exposed?

> Calling begin/end_multicaret_edit() here seems a little odd to me, as adding / removing carets shouldn't be changing any text? Looking at add_caret_at_carets in particular, it calls merge_overlapping_carets at the end which should achieve the same purpose?

I made add_caret always able to return a caret if in a multicaret edit, and it will be merged later. I needed to be able to add a caret even if it overlaps, so I can set the selection so it can get merged properly.

A better example would probably be `CodeEdit::toggle_foldable_line_for_all_carets()` which doesn't change text but needs to collapse carets without immediately merging them, and not be part of undo / redo.

kitbdev commented:
Fixed failing check and added collapse tests.

Paulb23 commented:
Appreciate the split.

> I wasn't able to expose get_line_ranges_from_carets, since I got compiler errors, I guess Vector<Point2i> can't be exposed?

Yeah can't expose `Vector`, have convert it to a `TypedArray` to do that.

> A better example would probably be CodeEdit::toggle_foldable_line_for_all_carets() which doesn't change text but needs to collapse carets without immediately merging them, and not be part of undo / redo.

I see, that makes sense. I guess it might be worth adding a note to the docs that this doesn't replace complex op calls but rather an addition. Though in this case an example might be clearer?

kitbdev commented:
>Yeah can't expose Vector, have convert it to a TypedArray to do that.

If I use a TypedArray, I think I won't be able to use ranged for each loops with it.
Also, it looks like there is no PackedVector2iArray for some reason.

> I see, that makes sense. I guess it might be worth adding a note to the docs that this doesn't replace complex op calls but rather an addition. Though in this case an example might be clearer?

I guess I thought it would be clear they are unrelated, but an example would be useful anyway.

Something like:
```gdscript
begin_complex_operation()
begin_multicaret_edit()
for i in range(get_caret_count()):
	if multicaret_edit_ignore_caret(i):
		continue
	# Logic here.
end_multicaret_edit()
end_complex_operation()
```

kitbdev commented:
Rebased to fix minor merge conflicts and update to new deprecated doc style.
Also, I think I can split off a few more things into separate PRs so its easier to review.

MewPurPur commented:
I think this PR is clear enough as is. I gave it a big 30min round of testing, trying out every functionality I could think of, I couldn't get any bugs to happen.

kitbdev commented:
Rebased to fix conflicts (#88474).
I removed the save multiple carets state and the goto line changes, since I wanted to make some changes to them and they can be done separately. I'll make a new PR for them.

Mickeon commented:
Going to chime in and say that this would also fix https://github.com/godotengine/godot/pull/86978 . What an insane list of bug reports

kitbdev commented:
> Going to chime in and say that this would also fix https://github.com/godotengine/godot/issues/76093

This doesn't actually fix https://github.com/godotengine/godot/issues/76093.
This PR should fix all other issues that were fixed in https://github.com/godotengine/godot/pull/73580, except for that one.
I didn't fix https://github.com/godotengine/godot/issues/76093 because it seemed like a bigger issue since it also affects word left/right behavior and may affect `TS->shaped_text_get_word_breaks`, but I didn't really look into it.

Paulb23 reviewed (commented):
Got around to finishing taking a look a the the first commit, some comments left inline.

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -4561,44 +4491,67 @@ bool TextEdit::is_multiple_carets_enabled() const {
+	}
+}
+
+int TextEdit::get_caret_count() const {
```
The functions should be in the correct order in the second commit, I moved this one since it made the git diff a little cleaner.

kitbdev commented on scene/gui/text_edit.h:
```diff
@@ -883,20 +888,26 @@ class TextEdit : public Control {
 
 	String get_selected_text(int p_caret = -1);
+	int get_selection_at(int p_line, int p_column, bool p_include_edges = true) const;
+	Vector<Point2i> get_line_ranges_from_carets(bool p_only_selections = false, bool p_merge_adjacent = true) const;
```
I'll add a wrapper for now. We can adjust it in the future after either https://github.com/godotengine/godot-proposals/issues/5486 or TypedArray range iterators.

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -1680,151 +1697,119 @@ void TextEdit::gui_input(const Ref<InputEvent> &p_gui_input) {
 						// Remove caret at clicked location.
-						if (carets.size() > 1) {
-							for (int i = 0; i < carets.size(); i++) {
+						if (get_caret_count() > 1) {
```
They are slightly different, this check includes non-selection carets, but the `mouse_over_selection_caret` only includes selections.

kitbdev replied:
This now makes a call to `get_selection_at_line_column` with the new arg.

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -5002,19 +4957,8 @@ bool TextEdit::is_drag_and_drop_selection_enabled() const {
 void TextEdit::set_selection_mode(SelectionMode p_mode, int p_line, int p_column, int p_caret) {
-	ERR_FAIL_INDEX(p_caret, carets.size());
-
+	// Parameters kept for compatibility.
```
I'll update this to use an actual compatibility method.

kitbdev commented on scene/gui/text_edit.h:
```diff
@@ -883,20 +888,26 @@ class TextEdit : public Control {
 	bool has_selection(int p_caret = -1) const;
 
 	String get_selected_text(int p_caret = -1);
+	int get_selection_at(int p_line, int p_column, bool p_include_edges = true) const;
```
I don't think there are any other methods that accept line and column as Point2i, only ones that return it.
Methods like `add_caret` take them separately.
Also would it be x: line, y: column or vice-versa?

kitbdev commented on scene/gui/text_edit.h:
```diff
@@ -452,18 +450,23 @@ class TextEdit : public Control {
 	void _pre_shift_selection(int p_caret);
-	void _post_shift_selection(int p_caret);
+
+	bool _selection_contains(int p_caret, int p_line, int p_column, bool p_include_edges = true, bool p_only_selections = true) const;
```
I could invert it to `p_include_non_selections`? I think `p_only_selections` sounds more clear, though.

Paulb23 replied:
Thinking about this with the other comment around `mouse_over_selection_caret`, wondering if we should merge them into a single method?

kitbdev replied:
I can give `get_selection_at_line_column()` (mouse_over_selection_caret) a `p_only_selections` argument and then it can be used instead of looping in `gui_input()` and `add_caret()`, but I can't remove all uses of `_selection_contains()`, since it is used in `add_caret_at_carets()`.

Paulb23 replied:
For `add_caret_at_carets` would that be due to the them overlapping?
Otherwise that sounds fine to me.

kitbdev replied:
Yes, it's to check if they overlap, so that the correct caret will be removed when they merge.

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -4862,39 +4805,48 @@ void TextEdit::set_caret_line(int p_line, bool p_adjust_viewport, bool p_can_be_
-		if (n_col >= row_end_col) {
-			n_col -= 1;
+	int n_col;
+	if (p_wrap_index >= 0) {
```
Updated in the second commit.
Sorry, I didn't split the doc changes properly. I must have merged something wrong in git.

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -5125,57 +5065,184 @@ void TextEdit::add_selection_for_next_occurrence() {
 		}
 	}
+
+	bool caret_moved = get_selection_origin_line(p_caret) != p_line;
```
Maybe, but I think I can simplify it a bit instead. Using `text.get_line_wrap_ranges()` should be better than using `get_line_wrapped_text()` and iterating over it.

kitbdev commented on scene/gui/text_edit.h:
```diff
@@ -883,20 +888,26 @@ class TextEdit : public Control {
 
-	int get_selection_line(int p_caret = 0) const;
-	int get_selection_column(int p_caret = 0) const;
+	void set_selection_origin_line(int p_line, int p_caret = 0, bool p_can_be_hidden = true, int p_wrap_index = -1);
```
This does make it more awkward, since the `p_caret` param is always used, and the other two are only sometimes used. With `p_caret` at the end they wouldn't even need to be optional.
But I guess if consistency is more important, I'll change it.

kitbdev commented:
Rebased and fixed minor conflicts after https://github.com/godotengine/godot/pull/85325

Mickeon reviewed (commented):
There's _several_ methods I feel like should not be exposed as they're a bit _too_ specific, and would require extra maintenance if something huge like this PR ever needs to happen again.

I'm only one person, though.

kitbdev commented on doc/classes/CodeEdit.xml:
```diff
@@ -144,6 +144,12 @@
 				Code regions are delimited using start and end tags (respectively [code]region[/code] and [code]endregion[/code] by default) preceded by one line comment delimiter. (eg. [code]#region[/code] and [code]#endregion[/code])
 			</description>
 		</method>
+		<method name="delete_lines">
+			<return type="void" />
+			<description>
+				Deletes all lines that are selected or have a caret on them.
+			</description>
+		</method>
```
I feel like these four methods are common enough to be exposed for CodeEdit, since they're common features in many IDEs. But, I can unexpose them if we want to wait before commiting to them.
Also, it seems that @Paulb23 wants all public methods to be exposed https://github.com/godotengine/godot/pull/86978#issuecomment-1904834139

kitbdev commented on doc/classes/TextEdit.xml:
```diff
@@ -83,15 +83,14 @@
 			<param index="3" name="to_line" type="int" />
 			<param index="4" name="to_col" type="int" />
 			<description>
-				Reposition the carets affected by the edit. This assumes edits are applied in edit order, see [method get_caret_index_edit_order].
 			</description>
```
`adjust_carets_after_edit` no longer does anything, and shouldn't do anything, since its behavior is essentially part of the other edit methods, like `remove_text()`.

Mickeon replied:
If it does nothing the description needs to be updated to explicitly state it ("This method does nothing.") for the same reasons as above.

kitbdev replied:
I can also restore the descriptions I removed of get_selection_line/column.

kitbdev commented on doc/classes/TextEdit.xml:
```diff
@@ -205,10 +241,9 @@
+		<method name="get_caret_index_edit_order" deprecated="Carets no longer need to be edited in any specific order. If the carets need to be sorted, use [method get_sorted_carets] instead.">
 			<return type="PackedInt32Array" />
 			<description>
-				Returns a list of caret indexes in their edit order, this done from bottom to top. Edit order refers to the way actions such as [method insert_text_at_caret] are applied.
 			</description>
```
`get_caret_index_edit_order` does still work, but the description is outdated since there is no more required edit order of carets.
Should I keep the description as it was or update it somehow?

Mickeon reviewed (approved):
This looks good on the class reference side. It all looks extremely promising. I hope this approval brings more attention to it.

MewPurPur reviewed (approved):
After a solid amount of testing, I still can't break it in any way. I even found a good amount of unreported bugs fixed by it, like drag-and-drop and error navigation breaking when you have selections in the document.

kitbdev commented on scene/gui/text_edit.h:
```diff
@@ -768,11 +776,13 @@ class TextEdit : public Control {
 	int get_first_non_whitespace_column(int p_line) const;
 
-	void swap_lines(int p_from_line, int p_to_line);
+	void swap_lines(int p_from_line, int p_to_line, bool p_swap_carets = true);
```
I think the compat methods should have actually used false for these parameters to have the functionality be more similar to the previous behavior.
The functionality is still a bit different though, so I can remove them.

kitbdev commented on scene/gui/text_edit.h:
```diff
@@ -768,11 +776,13 @@ class TextEdit : public Control {
+	void remove_line_at(int p_line, bool p_move_carets_down = true);
 
+	void insert_text_at_caret(const String &p_text, int p_caret = -1);
+	void insert_text(const String &p_text, int p_line, int p_column, bool p_before_selection_begin = true, bool p_before_selection_end = false);
```
I think `p_before_selection_begin/end` would be easier to understand for users in the context of inserting text than `p_include_selection_begin/end` would be.
`_offset_carets_after` doesn't have that context of inserting text and `p_include_selection_begin/end` are more accurate for how it works when offseting carets.
So, I think we shouldn't make them the same.

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -4963,7 +5057,7 @@ String TextEdit::get_word_under_caret(int p_caret) const {
 			if (words[i] <= get_caret_column(c) && words[i + 1] > get_caret_column(c)) {
 				selected_text += text[get_caret_line(c)].substr(words[i], words[i + 1] - words[i]);
-				if (p_caret == -1 && c != carets.size() - 1) {
+				if (p_caret == -1 && c != 0) {
```
I can add a simple test to check for this.

kitbdev commented:
Rebased to fix conflicts.
Removed args for `set_line` and `swap_lines` so they don't break compat, and other suggestions.

kitbdev commented:
rebased to fix merge conflicts

Paulb23 reviewed (approved):
Looks good to me, can fix any further issues once merged

Really appreciate the effort put into the unit tests as well :)

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/89850d553eeb259e208d0c577cd7bc1eabd3a90a), it works as expected. (I've tried things such as <kbd>Ctrl + D</kbd>, <kbd>Ctrl + Alt + D</kbd>, drag-and-drop with multiple selections and so on.)

akien-mga commented:
Great work! Thanks for putting up with the exhaustive review process :)
